From 392382a44c27135bfd0b02bf441b4f9582324bd7 Mon Sep 17 00:00:00 2001
From: P Praneesh <quic_ppranees@quicinc.com>
Date: Fri, 19 May 2023 14:12:33 +0530
Subject: [PATCH] ath12k: fix memory leak in WMI firmware stats

Memory allocated for firmware pdev, vdev and beacon statistics
are not released during rmmod. Fix it by calling ath12k_fw_stats_free
before hardware unregister. Also, while processing the firmware
stats received in the WMI event, the local list is getting spliced
and reinitialised. Avoid calling ath12k_fw_stats_free because there
are no elements in the list after splicing.

Signed-off-by: P Praneesh <quic_ppranees@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/dp.c  | 1 +
 drivers/net/wireless/ath/ath12k/mac.c | 1 +
 drivers/net/wireless/ath/ath12k/wmi.c | 5 +++++
 3 files changed, 7 insertions(+)

diff --git a/drivers/net/wireless/ath/ath12k/dp.c b/drivers/net/wireless/ath/ath12k/dp.c
index 5eb322d..50985b0 100644
--- a/drivers/net/wireless/ath/ath12k/dp.c
+++ b/drivers/net/wireless/ath/ath12k/dp.c
@@ -1169,6 +1169,7 @@ void ath12k_dp_pdev_free(struct ath12k_base *ab)
 		ar = ab->pdevs[i].ar;
 		ath12k_dp_rx_pdev_free(ab, i);
 		ath12k_dp_rx_pdev_mon_detach(ab, i);
+		ath12k_fw_stats_free(&ar->fw_stats);
 		ath12k_debugfs_unregister(ar);
 	}
 }
diff --git a/drivers/net/wireless/ath/ath12k/mac.c b/drivers/net/wireless/ath/ath12k/mac.c
index beab1c7..88d804d 100644
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -15146,6 +15146,7 @@ static int ath12k_mac_hw_register(struct ath12k_hw *ah)
 err_unregister_hw:
 	ar = ah->radio;
 	for (j = 0; j < i; j++) {
+		ath12k_fw_stats_free(&ar->fw_stats);
 		ath12k_debugfs_unregister(ar);
 		ar++;
 	}
diff --git a/drivers/net/wireless/ath/ath12k/wmi.c b/drivers/net/wireless/ath/ath12k/wmi.c
index 5921ab2..2bd0c7c 100644
--- a/drivers/net/wireless/ath/ath12k/wmi.c
+++ b/drivers/net/wireless/ath/ath12k/wmi.c
@@ -11532,6 +11532,11 @@ complete:
 	rcu_read_unlock();
 	spin_unlock_bh(&ar->data_lock);
 
+	/* Since the stats's pdev, vdev and beacon list are spliced and reinitialised
+	 * at this point, no need to free the individual list.
+	 */
+	return;
+
 free:
 	ath12k_fw_stats_free(&stats);
 }
-- 
2.17.1

